From 02349056c282391e5540b42b7c0c4dcd341e40bb Mon Sep 17 00:00:00 2001
From: Jonathan <jonathan.lvaldivieso@upm.es>
Date: Tue, 14 Apr 2026 16:51:24 +0200
Subject: [PATCH] net: ethernet: spi: add LowRISC Ethernet and OpenTitan QSPI
 drivers

Add platform drivers required by the Cheshire CVA6/RISC-V SoC:

- LowRISC 100MHz Ethernet controller (drivers/net/ethernet/lowrisc/)
  with MDIO bitbang PHY management and NAPI receive polling.

- OpenTitan QSPI SPI host controller (drivers/spi/spi-opentitan.c)
  used for SD card and NOR flash access on the Cheshire platform.

Both drivers have been ported from kernel 5.10 to 6.1.120:
  - netif_napi_add: removed weight parameter (4 to 3 args)
  - dev_addr: use eth_hw_addr_set() for const dev_addr
  - lowrisc_update_address: accept const u8 pointer
  - ndo_do_ioctl renamed to ndo_eth_ioctl
  - platform_driver .remove changed to .remove_new with void return
---
 drivers/net/ethernet/Kconfig                  |   1 +
 drivers/net/ethernet/Makefile                 |   1 +
 drivers/net/ethernet/lowrisc/Kconfig          |  26 +
 drivers/net/ethernet/lowrisc/Makefile         |   6 +
 drivers/net/ethernet/lowrisc/lowrisc_100MHz.c | 830 ++++++++++++++++++
 drivers/net/ethernet/lowrisc/lowrisc_100MHz.h |  55 ++
 drivers/spi/Kconfig                           |   5 +
 drivers/spi/Makefile                          |   1 +
 drivers/spi/spi-opentitan.c                   | 755 ++++++++++++++++
 9 files changed, 1680 insertions(+)
 create mode 100644 drivers/net/ethernet/lowrisc/Kconfig
 create mode 100644 drivers/net/ethernet/lowrisc/Makefile
 create mode 100644 drivers/net/ethernet/lowrisc/lowrisc_100MHz.c
 create mode 100644 drivers/net/ethernet/lowrisc/lowrisc_100MHz.h
 create mode 100644 drivers/spi/spi-opentitan.c

diff --git a/drivers/net/ethernet/Kconfig b/drivers/net/ethernet/Kconfig
index 5a274b99f299..21b1b513d459 100644
--- a/drivers/net/ethernet/Kconfig
+++ b/drivers/net/ethernet/Kconfig
@@ -122,6 +122,7 @@ config LANTIQ_XRX200
 
 source "drivers/net/ethernet/adi/Kconfig"
 source "drivers/net/ethernet/litex/Kconfig"
+source "drivers/net/ethernet/lowrisc/Kconfig"
 source "drivers/net/ethernet/marvell/Kconfig"
 source "drivers/net/ethernet/mediatek/Kconfig"
 source "drivers/net/ethernet/mellanox/Kconfig"
diff --git a/drivers/net/ethernet/Makefile b/drivers/net/ethernet/Makefile
index 0d872d4efcd1..f16c64974ba6 100644
--- a/drivers/net/ethernet/Makefile
+++ b/drivers/net/ethernet/Makefile
@@ -56,6 +56,7 @@ obj-$(CONFIG_KORINA) += korina.o
 obj-$(CONFIG_LANTIQ_ETOP) += lantiq_etop.o
 obj-$(CONFIG_LANTIQ_XRX200) += lantiq_xrx200.o
 obj-$(CONFIG_NET_VENDOR_LITEX) += litex/
+obj-$(CONFIG_NET_VENDOR_LOWRISC) += lowrisc/
 obj-$(CONFIG_NET_VENDOR_MARVELL) += marvell/
 obj-$(CONFIG_NET_VENDOR_MEDIATEK) += mediatek/
 obj-$(CONFIG_NET_VENDOR_MELLANOX) += mellanox/
diff --git a/drivers/net/ethernet/lowrisc/Kconfig b/drivers/net/ethernet/lowrisc/Kconfig
new file mode 100644
index 000000000000..ab95f70ab9d8
--- /dev/null
+++ b/drivers/net/ethernet/lowrisc/Kconfig
@@ -0,0 +1,26 @@
+#
+# Xilink device configuration
+#
+
+config NET_VENDOR_LOWRISC
+	bool "Lowrisc devices"
+	default y
+	depends on RISCV
+	help
+	  If you have a network (Ethernet) card belonging to this class, say Y.
+
+	  Note that the answer to this question doesn't directly affect the
+	  kernel: saying N will just cause the configurator to skip all
+	  the questions about Lowrisc devices. If you say Y, you will be asked
+	  for your specific card in the following questions.
+
+if NET_VENDOR_LOWRISC
+
+config LOWRISC_DIGILENT_100MHZ
+	tristate "Lowrisc 100MHz Ethernet Nexys4_DDR support"
+	depends on RISCV
+	select PHYLIB
+	help
+	  This driver supports the 100MHz Ethernet for Nexys4_DDR Digilent boards from Lowrisc.
+
+endif # NET_VENDOR_LOWRISC
diff --git a/drivers/net/ethernet/lowrisc/Makefile b/drivers/net/ethernet/lowrisc/Makefile
new file mode 100644
index 000000000000..4eb3adee161d
--- /dev/null
+++ b/drivers/net/ethernet/lowrisc/Makefile
@@ -0,0 +1,6 @@
+#
+# Makefile for the Lowrisc network device driver.
+#
+
+obj-$(CONFIG_LOWRISC_DIGILENT_100MHZ) += lowrisc_100MHz.o
+CFLAGS_lowrisc_100MHz.o := -DDEBUG
diff --git a/drivers/net/ethernet/lowrisc/lowrisc_100MHz.c b/drivers/net/ethernet/lowrisc/lowrisc_100MHz.c
new file mode 100644
index 000000000000..e2d3ec1ec67b
--- /dev/null
+++ b/drivers/net/ethernet/lowrisc/lowrisc_100MHz.c
@@ -0,0 +1,830 @@
+/*
+ * Lowrisc Ether100MHz Linux driver for the Lowrisc Ethernet 100MHz device.
+ *
+ * This is an experimental driver which is based on the original emac_lite
+ * driver from John Williams <john.williams@xilinx.com>.
+ *
+ * 2007 - 2013 (c) Xilinx, Inc.
+ * PHY control portions copyright (C) 2015 Microchip Technology
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/mii.h>
+#include <linux/ethtool.h>
+#include <linux/phy.h>
+#include <linux/module.h>
+#include <linux/uaccess.h>
+#include <linux/netdevice.h>
+#include <linux/etherdevice.h>
+#include <linux/skbuff.h>
+#include <linux/io.h>
+#include <linux/slab.h>
+#include <linux/of_address.h>
+#include <linux/of_device.h>
+#include <linux/of_platform.h>
+#include <linux/of_mdio.h>
+#include <linux/of_net.h>
+#include <linux/phy.h>
+#include <linux/interrupt.h>
+#include <linux/timer.h>
+#include <linux/spinlock.h>
+#include <linux/gpio.h>
+#include <linux/slab.h>
+#include <linux/delay.h>
+#include <linux/platform_device.h>
+#include <linux/mdio-bitbang.h>
+#include <linux/mdio-gpio.h>
+#include "lowrisc_100MHz.h"
+
+#define DRIVER_AUTHOR	"WOOJUNG HUH <woojung.huh@microchip.com>"
+#define DRIVER_DESC	"Microchip LAN8720 PHY driver"
+#define DRIVER_NAME     "lowrisc-eth"
+
+/* General Ethernet Definitions */
+#define XEL_ARP_PACKET_SIZE		28	/* Max ARP packet size */
+#define XEL_HEADER_IP_LENGTH_OFFSET	16	/* IP Length Offset */
+
+#define TX_TIMEOUT		(60*HZ)		/* Tx timeout is 60 seconds. */
+
+/**
+ * struct net_local - Our private per device data
+ * @ndev:		instance of the network device
+ * @reset_lock:		lock used for synchronization
+ * @phy_dev:		pointer to the PHY device
+ * @phy_node:		pointer to the PHY device node
+ * @mii_bus:		pointer to the MII bus
+ * @last_link:		last link status
+ */
+struct net_local {
+  struct mdiobb_ctrl ctrl; /* must be first for bitbang driver to work */
+  void __iomem *ioaddr;
+  struct net_device *ndev;
+  u32 msg_enable;
+  
+  struct phy_device *phy_dev;
+  struct mii_bus *mii_bus;
+  int last_duplex;
+  int last_carrier;
+  
+  /* Spinlock */
+  struct mutex lock;
+  uint32_t last_mdio_gpio;
+  int irq, num_tests, phy_addr;
+
+  struct napi_struct napi;
+};
+
+static void inline eth_write(struct net_local *priv, size_t addr, int data)
+{
+  volatile uint64_t *eth_base = (volatile uint64_t *)(priv->ioaddr);
+  eth_base[addr >> 3] = data;
+}
+
+static void inline eth_copyout(struct net_local *priv, uint8_t *data, int len)
+{
+  int i, rnd = ((len-1)|7)+1;
+  volatile uint64_t *eth_base = (volatile uint64_t *)(priv->ioaddr);
+  if (!(((size_t)data) & 7))
+    {
+      uint64_t *ptr = (uint64_t *)data;
+      for (i = 0; i < rnd/8; i++)
+        eth_base[TXBUFF_OFFSET/8 + i] = ptr[i];
+    }
+  else // We can't unfortunately rely on the skb being word aligned
+    {
+      uint64_t notptr;
+      for (i = 0; i < rnd/8; i++)
+        {
+          memcpy(&notptr, data+(i<<3), sizeof(uint64_t));
+          eth_base[TXBUFF_OFFSET/8 + i] = notptr;
+        }
+    }
+}
+
+static volatile inline int eth_read(struct net_local *priv, size_t addr)
+{
+  volatile uint64_t *eth_base = (volatile uint64_t *)(priv->ioaddr);
+  return eth_base[addr >> 3];
+}
+
+static inline void eth_copyin(struct net_local *priv, uint8_t *data, int len, int start)
+{
+  int i, rnd = ((len-1)|7)+1;
+  volatile uint64_t *eth_base = (volatile uint64_t *)(priv->ioaddr);
+  if (!(((size_t)data) & 7))
+    {
+      uint64_t *ptr = (uint64_t *)data;
+      for (i = 0; i < rnd/8; i++)
+        ptr[i] = eth_base[start + i];
+    }
+  else // We can't unfortunately rely on the skb being word aligned
+    {
+      for (i = 0; i < rnd/8; i++)
+        {
+          uint64_t notptr = eth_base[start + i];
+          memcpy(data+(i<<3), &notptr, sizeof(uint64_t));
+        }
+    }
+}
+
+static void inline eth_enable_irq(struct net_local *priv)
+{
+  volatile uint64_t *eth_base = (volatile uint64_t *)(priv->ioaddr);
+  eth_base[MACHI_OFFSET >> 3] |= MACHI_IRQ_EN;
+  mmiowb();
+}
+
+static void inline eth_disable_irq(struct net_local *priv)
+{
+  volatile uint64_t *eth_base = (volatile uint64_t *)(priv->ioaddr);
+  eth_base[MACHI_OFFSET >> 3] &= ~MACHI_IRQ_EN;
+  mmiowb();
+}
+
+/**
+ * lowrisc_update_address - Update the MAC address in the device
+ * @drvdata:	Pointer to the Ether100MHz device private data
+ * @address_ptr:Pointer to the MAC address (MAC address is a 48-bit value)
+ *
+ * Tx must be idle and Rx should be idle for deterministic results.
+ * It is recommended that this function should be called after the
+ * initialization and before transmission of any packets from the device.
+ * The MAC address can be programmed using any of the two transmit
+ * buffers (if configured).
+ */
+
+static void lowrisc_update_address(struct net_local *priv, const u8 *address_ptr)
+{
+  uint32_t macaddr_lo, macaddr_hi;
+  memcpy (&macaddr_lo, address_ptr+2, sizeof(uint32_t));
+  memcpy (&macaddr_hi, address_ptr+0, sizeof(uint16_t));
+  mutex_lock(&priv->lock);
+  eth_write(priv, MACLO_OFFSET, htonl(macaddr_lo));
+  eth_write(priv, MACHI_OFFSET, htons(macaddr_hi));
+  eth_write(priv, RFCS_OFFSET, 8); /* use 8 buffers */
+  mutex_unlock(&priv->lock);
+}
+
+/**
+ * lowrisc_read_mac_address - Read the MAC address in the device
+ * @drvdata:	Pointer to the Ether100MHz device private data
+ * @address_ptr:Pointer to the 6-byte buffer to receive the MAC address (MAC address is a 48-bit value)
+ *
+ * In lowrisc the starting value is programmed by the boot loader according to DIP switch [15:12]
+ */
+
+static void lowrisc_read_mac_address(struct net_local *priv, u8 *address_ptr)
+{
+  uint32_t macaddr_hi, macaddr_lo;
+  mutex_lock(&priv->lock);
+  macaddr_hi = ntohs(eth_read(priv, MACHI_OFFSET)&MACHI_MACADDR_MASK);
+  macaddr_lo = ntohl(eth_read(priv, MACLO_OFFSET));
+  mutex_unlock(&priv->lock);
+  memcpy (address_ptr+2, &macaddr_lo, sizeof(uint32_t));
+  memcpy (address_ptr+0, &macaddr_hi, sizeof(uint16_t));
+}
+
+/**
+ * lowrisc_set_mac_address - Set the MAC address for this device
+ * @dev:	Pointer to the network device instance
+ * @addr:	Void pointer to the sockaddr structure
+ *
+ * This function copies the HW address from the sockaddr strucutre to the
+ * net_device structure and updates the address in HW.
+ *
+ * Return:	Error if the net device is busy or 0 if the addr is set
+ *		successfully
+ */
+static int lowrisc_set_mac_address(struct net_device *ndev, void *address)
+{
+	struct net_local *priv = netdev_priv(ndev);
+	struct sockaddr *addr = address;
+	eth_hw_addr_set(ndev, addr->sa_data);
+	lowrisc_update_address(priv, ndev->dev_addr);
+	return 0;
+}
+
+/**
+ * lowrisc_tx_timeout - Callback for Tx Timeout
+ * @dev:	Pointer to the network device
+ *
+ * This function is called when Tx time out occurs for Ether100MHz device.
+ */
+static void lowrisc_tx_timeout(struct net_device *ndev, unsigned int txqueue)
+{
+	struct net_local *priv = netdev_priv(ndev);
+
+	dev_err(&priv->ndev->dev, "Exceeded transmit timeout of %lu ms\n",
+		TX_TIMEOUT * 1000UL / HZ);
+
+	ndev->stats.tx_errors++;
+
+	/* Reset the device */
+	mutex_lock(&priv->lock);
+
+	/* Shouldn't really be necessary, but shouldn't hurt */
+	netif_stop_queue(ndev);
+
+	/* To exclude tx timeout */
+        netif_trans_update(ndev); /* prevent tx timeout */
+
+	/* We're all ready to go. Start the queue */
+	netif_wake_queue(ndev);
+	mutex_unlock(&priv->lock);
+}
+
+/**
+ * lowrisc_close - Close the network device
+ * @dev:	Pointer to the network device
+ *
+ * This function stops the Tx queue, disables interrupts and frees the IRQ for
+ * the Ether100MHz device.
+ * It also disconnects the phy device associated with the Ether100MHz device.
+ */
+static int lowrisc_close(struct net_device *ndev)
+{
+	struct net_local *priv = netdev_priv(ndev);
+
+	netif_stop_queue(ndev);
+        napi_disable(&priv->napi);
+        mutex_lock(&priv->lock);
+	eth_disable_irq(priv);
+        mutex_unlock(&priv->lock);
+	if (priv->irq)
+		free_irq(priv->irq, priv);
+	priv->irq = 0;
+        printk("Close device, free interrupt\n");
+        
+	if (priv->phy_dev)
+		phy_disconnect(priv->phy_dev);
+	priv->phy_dev = NULL;
+
+	return 0;
+}
+
+/**
+ * lowrisc_remove_ndev - Free the network device
+ * @ndev:	Pointer to the network device to be freed
+ *
+ * This function un maps the IO region of the Ether100MHz device and frees the net
+ * device.
+ */
+static void lowrisc_remove_ndev(struct net_device *ndev)
+{
+	if (ndev) {
+		free_netdev(ndev);
+	}
+}
+
+static void lowrisc_phy_adjust_link(struct net_device *ndev)
+{
+	struct net_local *priv = netdev_priv(ndev);
+	struct phy_device *phy_dev = priv->phy_dev;
+	int carrier;
+
+	if (phy_dev->duplex != priv->last_duplex) {
+		if (phy_dev->duplex) {
+			netif_dbg(priv, link, priv->ndev, "full duplex mode\n");
+		} else {
+			netif_dbg(priv, link, priv->ndev, "half duplex mode\n");
+		}
+
+		priv->last_duplex = phy_dev->duplex;
+	}
+
+	carrier = netif_carrier_ok(ndev);
+	if (carrier != priv->last_carrier) {
+		if (carrier)
+			netif_dbg(priv, link, priv->ndev, "carrier OK\n");
+		else
+			netif_dbg(priv, link, priv->ndev, "no carrier\n");
+		priv->last_carrier = carrier;
+	}
+}
+
+static int lowrisc_mii_probe(struct net_device *ndev)
+{
+	struct net_local *priv = netdev_priv(ndev);
+	struct phy_device *phydev = NULL;
+	const char *phyname;
+
+	BUG_ON(priv->phy_dev);
+
+	/* find the first (lowest address) PHY
+	 * on the current MAC's MII bus
+	 */
+	for (priv->phy_addr = 1; priv->phy_addr < PHY_MAX_ADDR; priv->phy_addr++)
+          {
+            phydev = mdiobus_get_phy(priv->mii_bus, priv->phy_addr);
+            if (phydev) {
+              /* break out with first one found */
+              break;
+            }
+          }
+	if (!phydev) {
+          netdev_err(ndev, "no PHY found in range address 0..%d\n", PHY_MAX_ADDR-1);
+		return -ENODEV;
+	}
+
+	phyname = phydev_name(phydev);
+	printk("Probing %s (address %d)\n", phyname, priv->phy_addr);
+	
+	phydev = phy_connect(ndev, phyname,
+			     lowrisc_phy_adjust_link, PHY_INTERFACE_MODE_MII);
+
+	if (IS_ERR(phydev)) {
+		netdev_err(ndev, "Could not attach to PHY\n");
+		return PTR_ERR(phydev);
+	}
+
+	/* mask with MAC supported features */
+	linkmode_copy(phydev->advertising, phydev->supported);
+
+	phy_attached_info(phydev);
+
+	priv->phy_dev = phydev;
+	priv->last_duplex = -1;
+	priv->last_carrier = -1;
+
+	return 0;
+}
+
+static void mdio_dir(struct mdiobb_ctrl *ctrl, int dir)
+{
+  struct net_local *priv = (struct net_local *)ctrl; /* struct mdiobb_ctrl must be first in net_local for bitbang driver to work */
+  volatile uint64_t *eth_base = (volatile uint64_t *)(priv->ioaddr);
+  if (dir)
+    priv->last_mdio_gpio |= MDIOCTRL_MDIOOEN_MASK; // output driving
+  else
+    priv->last_mdio_gpio &= ~MDIOCTRL_MDIOOEN_MASK; // input receiving
+  eth_base[MDIOCTRL_OFFSET >> 3] = priv->last_mdio_gpio;
+  mmiowb();
+}
+
+static int mdio_get(struct mdiobb_ctrl *ctrl)
+{
+  int retval;
+  struct net_local *priv = (struct net_local *)ctrl; /* struct mdiobb_ctrl must be first in net_local for bitbang driver to work */
+  volatile uint64_t *eth_base = (volatile uint64_t *)(priv->ioaddr);
+  retval = eth_base[MDIOCTRL_OFFSET >> 3];
+  return retval & MDIOCTRL_MDIOIN_MASK ? 1:0;
+}
+
+static void mdio_set(struct mdiobb_ctrl *ctrl, int what)
+{
+  struct net_local *priv = (struct net_local *)ctrl; /* struct mdiobb_ctrl must be first in net_local for bitbang driver to work */
+  volatile uint64_t *eth_base = (volatile uint64_t *)(priv->ioaddr);
+  if (what)
+    priv->last_mdio_gpio |= MDIOCTRL_MDIOOUT_MASK;
+  else
+    priv->last_mdio_gpio &= ~MDIOCTRL_MDIOOUT_MASK;
+  eth_base[MDIOCTRL_OFFSET >> 3] = priv->last_mdio_gpio;
+  mmiowb();
+}
+
+static void mdc_set(struct mdiobb_ctrl *ctrl, int what)
+{
+  struct net_local *priv = (struct net_local *)ctrl; /* struct mdiobb_ctrl must be first in net_local for bitbang driver to work */
+  volatile uint64_t *eth_base = (volatile uint64_t *)(priv->ioaddr);
+  if (what)
+    priv->last_mdio_gpio |= MDIOCTRL_MDIOCLK_MASK;
+  else
+    priv->last_mdio_gpio &= ~MDIOCTRL_MDIOCLK_MASK;
+  eth_base[MDIOCTRL_OFFSET >> 3] = priv->last_mdio_gpio;
+  mmiowb();
+}
+
+static struct mdiobb_ops mdio_gpio_ops = {
+        .owner = THIS_MODULE,
+        .set_mdc = mdc_set,
+        .set_mdio_dir = mdio_dir,
+        .set_mdio_data = mdio_set,
+        .get_mdio_data = mdio_get,
+};
+
+static int lowrisc_mii_init(struct net_device *ndev)
+{
+        struct mii_bus *new_bus;
+	struct net_local *priv = netdev_priv(ndev);
+	int err = -ENXIO;
+	
+	priv->ctrl.ops = &mdio_gpio_ops;
+        new_bus = alloc_mdio_bitbang(&(priv->ctrl));
+
+	if (!new_bus) {
+		err = -ENOMEM;
+		goto err_out_1;
+	}
+	snprintf(new_bus->id, MII_BUS_ID_SIZE, "lowrisc-0");
+        new_bus->name = "GPIO Bitbanged LowRISC",
+
+        new_bus->phy_mask = ~(1 << 1);
+        new_bus->phy_ignore_ta_mask = 0;
+
+	mutex_init(&(new_bus->mdio_lock));
+	
+	priv->mii_bus = new_bus;
+	priv->mii_bus->priv = priv;
+
+	/* Mask all PHYs except ID 1 (internal) */
+	priv->mii_bus->phy_mask = ~(1 << 1);
+
+	if (mdiobus_register(priv->mii_bus)) {
+		netif_warn(priv, probe, priv->ndev, "Error registering mii bus\n");
+		goto err_out_free_bus_2;
+	}
+
+	if (lowrisc_mii_probe(ndev) < 0) {
+		netif_warn(priv, probe, priv->ndev, "Error probing mii bus\n");
+		goto err_out_unregister_bus_3;
+	}
+
+	return 0;
+
+err_out_unregister_bus_3:
+	mdiobus_unregister(priv->mii_bus);
+err_out_free_bus_2:
+	mdiobus_free(priv->mii_bus);
+err_out_1:
+	return err;
+}
+/**********************/
+/* Interrupt Handlers */
+/**********************/
+
+/**
+ * lowrisc_ether_isr - Interrupt handler for frames received
+ * @priv:	Pointer to the struct net_local
+ *
+ * This function allocates memory for a socket buffer, fills it with data
+ * received and hands it over to the TCP/IP stack.
+ */
+
+static int lowrisc_ether_poll(struct napi_struct *napiptr, int budget)
+{
+  int rsr, buf, rx_count = 0;
+  struct net_local *priv = container_of(napiptr, struct net_local, napi);
+  struct net_device *ndev = priv->ndev;
+  rsr = eth_read(priv, RSR_OFFSET);
+  buf = rsr & RSR_RECV_FIRST_MASK;
+  /* Check if there is Rx Data available */
+  while ((rsr & RSR_RECV_DONE_MASK) && (rx_count < budget))
+    {
+      int len = eth_read(priv, RPLR_OFFSET+((buf&7)<<3)) - 4; /* discard FCS bytes ?? */
+      if ((len >= 60) && (len <= ETH_FRAME_LEN + ETH_FCS_LEN))
+	{
+	  int rnd = ((len-1)|7)+1; /* round to a multiple of 8 */
+	  struct sk_buff *skb = __napi_alloc_skb(&priv->napi, rnd, GFP_ATOMIC|__GFP_NOWARN); // Don't warn, just drop surplus packets
+	  if (unlikely(!skb))
+	    {
+	      /* Couldn't get memory, we carry on regardless and drop if necessary */
+	      ndev->stats.rx_dropped++;
+	    }
+	  else
+	    {
+	      int start = RXBUFF_OFFSET/8 + ((buf&7)<<8);
+              skb_put(skb, len);	/* Tell the skb how much data we got */
+	      
+              eth_copyin(priv, skb->data, len, start);
+              skb->protocol = eth_type_trans(skb, ndev);
+              netif_receive_skb(skb);
+              ndev->stats.rx_packets++;
+              ndev->stats.rx_bytes += len;
+              ++rx_count;
+            }
+        }
+      else
+	  ndev->stats.rx_errors++;
+      /* acknowledge, even if an error occurs, to reset irq */
+      eth_write(priv, RSR_OFFSET, ++buf);
+      rsr = eth_read(priv, RSR_OFFSET);
+    }
+
+  if (rsr & RSR_RECV_DONE_MASK)
+    {
+      return 1;
+    }
+  else
+    {
+      napi_complete_done(&priv->napi, rx_count);
+      eth_enable_irq(priv);
+      return 0;
+    }
+}
+
+irqreturn_t lowrisc_ether_isr(int irq, void *priv_id)
+{
+  int rsr = 0;
+  struct net_local *priv = priv_id;
+  struct net_device *ndev = priv->ndev;
+  if (napi_schedule_prep(&priv->napi))
+    {
+      eth_disable_irq(priv);
+      __napi_schedule(&priv->napi);
+    }
+  else /* we are in denial of service mode */
+    do
+      {
+        int buf = rsr & RSR_RECV_FIRST_MASK;
+        ndev->stats.rx_dropped++;            
+        eth_write(priv, RSR_OFFSET, ++buf);
+        rsr = eth_read(priv, RSR_OFFSET);
+      }
+    while (rsr & RSR_RECV_DONE_MASK);
+  return IRQ_HANDLED;
+}
+
+static int lowrisc_get_regs_len(struct net_device __always_unused *netdev)
+{
+#define LOWRISC_REGS_LEN 64	/* overestimate */
+  return LOWRISC_REGS_LEN * sizeof(u32);
+}
+
+static void lowrisc_get_regs(struct net_device *ndev,
+			   struct ethtool_regs *regs, void *p)
+{
+  struct net_local *priv = netdev_priv(ndev);
+  struct phy_device *phy = priv->phy_dev;
+
+  u32 *regs_buff = p;
+  int i;
+
+  memset(p, 0, LOWRISC_REGS_LEN * sizeof(u32));
+
+  regs->version = 0;
+  mutex_lock(&priv->lock);
+  for (i = 0; i < LOWRISC_REGS_LEN; i++)
+    {
+      if (i >= 32)
+	regs_buff[i] = eth_read(priv, MACLO_OFFSET+((i-32)<<3));
+      else
+	{
+	regs_buff[i] = phy_read(phy, i);
+	}
+    }
+  mutex_unlock(&priv->lock);
+}
+
+static const struct ethtool_ops lowrisc_ethtool_ops = {
+	.get_regs_len		= lowrisc_get_regs_len,
+	.get_regs		= lowrisc_get_regs,
+        .get_link_ksettings     = phy_ethtool_get_link_ksettings,
+        .set_link_ksettings     = phy_ethtool_set_link_ksettings        
+};
+
+/**
+ * lowrisc_open - Open the network device
+ * @dev:	Pointer to the network device
+ *
+ * This function sets the MAC address, requests an IRQ and enables interrupts
+ * for the Ether100MHz device and starts the Tx queue.
+ * It also connects to the phy device, if MDIO is included in Ether100MHz device.
+ */
+
+static int lowrisc_open(struct net_device *ndev)
+{
+  int retval = 0;
+  struct net_local *priv = netdev_priv(ndev);
+  ndev->ethtool_ops = &lowrisc_ethtool_ops;
+
+  /* Set the MAC address each time opened */
+  lowrisc_update_address(priv, ndev->dev_addr);
+  
+  if (priv->phy_dev) {
+    linkmode_copy(priv->phy_dev->advertising, priv->phy_dev->supported);
+    phy_start(priv->phy_dev);
+  }
+  
+  /* Grab the IRQ */
+  printk("Open device, request interrupt %d\n", priv->irq);
+  retval = request_irq(priv->irq, lowrisc_ether_isr, 0, ndev->name, priv);
+  if (retval) {
+    dev_err(&priv->ndev->dev, "Could not allocate interrupt %d\n", priv->irq);
+    if (priv->phy_dev)
+      phy_disconnect(priv->phy_dev);
+    priv->phy_dev = NULL;
+    return retval;
+  }
+  
+  lowrisc_update_address(priv, ndev->dev_addr);
+
+  /* We're ready to go */
+  napi_enable(&priv->napi);
+  netif_start_queue(ndev);
+
+  /* enable the irq */
+  eth_enable_irq(priv);
+  return 0;
+}
+
+/**
+ * lowrisc_send - Transmit a frame
+ * @orig_skb:	Pointer to the socket buffer to be transmitted
+ * @dev:	Pointer to the network device
+ *
+ * This function checks if the Tx buffer of the Ether100MHz device is free to send
+ * data. If so, it fills the Tx buffer with data from socket buffer data,
+ * updates the stats and frees the socket buffer.
+ * Return:	0, always.
+ */
+static netdev_tx_t lowrisc_send(struct sk_buff *new_skb, struct net_device *ndev)
+{
+	struct net_local *priv = netdev_priv(ndev);
+	unsigned int len = new_skb->len;
+        int rslt;
+	if (mutex_is_locked(&priv->lock))
+          return NETDEV_TX_BUSY;
+        rslt = eth_read(priv, TPLR_OFFSET);
+        if (rslt & TPLR_BUSY_MASK)
+          return NETDEV_TX_BUSY;
+        eth_copyout(priv, new_skb->data, len);
+        eth_write(priv, TPLR_OFFSET, len);
+
+	skb_tx_timestamp(new_skb);
+
+	ndev->stats.tx_bytes += len;
+	ndev->stats.tx_packets++;
+	dev_consume_skb_any(new_skb);
+
+	return NETDEV_TX_OK;
+}
+
+static int lowrisc_mii_ioctl(struct net_device *netdev, struct ifreq *ifr, int cmd)
+{
+        struct net_local *priv = netdev_priv(netdev);
+	struct phy_device *phy = priv->phy_dev;
+        struct mii_ioctl_data *data = if_mii(ifr);
+
+        switch (cmd) {
+        case SIOCGMIIPHY:
+                data->phy_id = 1;
+                break;
+        case SIOCGMIIREG:
+                data->val_out = phy_read(phy, data->reg_num);
+                break;
+        case SIOCSMIIREG:
+                phy_write(phy, data->reg_num, data->val_in);
+                break;
+        default:
+                return -EOPNOTSUPP;
+        }
+        return 0;
+	}
+
+static struct net_device_stats *lowrisc_get_stats(struct net_device *ndev)
+{
+        return &ndev->stats;
+}
+
+static void lowrisc_set_rx_mode(struct net_device *ndev)
+{
+	struct net_local *priv = netdev_priv(ndev);
+        volatile uint64_t *eth_base = (volatile uint64_t *)(priv->ioaddr);
+
+        if (ndev->flags & IFF_PROMISC)
+          {
+            /* Set promiscuous. */
+            eth_base[MACHI_OFFSET >> 3] |= MACHI_ALLPKTS_MASK;
+          }
+        else
+          {
+            eth_base[MACHI_OFFSET >> 3] &= ~MACHI_ALLPKTS_MASK;            
+          }
+}
+static struct net_device_ops lowrisc_netdev_ops = {
+	.ndo_open		= lowrisc_open,
+	.ndo_stop		= lowrisc_close,
+	.ndo_start_xmit		= lowrisc_send,
+        .ndo_get_stats          = lowrisc_get_stats,
+        .ndo_set_rx_mode        = lowrisc_set_rx_mode,
+	.ndo_set_mac_address	= lowrisc_set_mac_address,
+	.ndo_tx_timeout		= lowrisc_tx_timeout,
+	.ndo_eth_ioctl          = lowrisc_mii_ioctl,
+};
+
+/**
+ * lowrisc_of_probe - Probe method for the Ether100MHz device.
+ * @ofdev:	Pointer to OF device structure
+ * @match:	Pointer to the structure used for matching a device
+ *
+ * This function probes for the Ether100MHz device in the device tree.
+ * It initializes the driver data structure and the hardware, sets the MAC
+ * address and registers the network device.
+ * It also registers a mii_bus for the Ether100MHz device, if MDIO is included
+ * in the device.
+ *
+ * Return:	0, if the driver is bound to the Ether100MHz device, or
+ *		a negative error if there is failure.
+ */
+static int lowrisc_100MHz_probe(struct platform_device *ofdev)
+{
+	struct net_device *ndev = NULL;
+	struct net_local *priv = NULL;
+	struct device *dev = &ofdev->dev;
+        struct resource *lowrisc_ethernet;
+	unsigned char mac_address[7];
+	int rc = 0;
+
+        lowrisc_ethernet = platform_get_resource(ofdev, IORESOURCE_MEM, 0);
+
+	/* Create an ethernet device instance */
+	ndev = alloc_etherdev(sizeof(struct net_local));
+	if (!ndev)
+		return -ENOMEM;
+
+	dev_set_drvdata(dev, ndev);
+	SET_NETDEV_DEV(ndev, &ofdev->dev);
+        platform_set_drvdata(ofdev, ndev);
+        
+	priv = netdev_priv(ndev);
+	priv->ndev = ndev;
+        priv->ioaddr = devm_ioremap_resource(&ofdev->dev, lowrisc_ethernet);
+	mutex_init(&priv->lock);
+
+        priv->num_tests = 1;
+        
+	ndev->netdev_ops = &lowrisc_netdev_ops;
+	ndev->flags &= ~IFF_MULTICAST;
+	ndev->watchdog_timeo = TX_TIMEOUT;
+        netif_napi_add(ndev, &priv->napi, lowrisc_ether_poll);
+
+	printk("lowrisc-digilent-ethernet: Lowrisc ethernet platform (%llX-%llX) mapped to %lx\n",
+               lowrisc_ethernet[0].start,
+               lowrisc_ethernet[0].end,
+               (size_t)(priv->ioaddr));
+
+        priv->irq = platform_get_irq(ofdev, 0);
+        
+        /* get the MAC address set by the boot loader */
+        lowrisc_read_mac_address(priv, mac_address);
+	eth_hw_addr_set(ndev, mac_address);
+
+	/* Set the MAC address in the Ether100MHz device */
+	lowrisc_update_address(priv, ndev->dev_addr);
+
+	lowrisc_mii_init(ndev);
+
+	/* Finally, register the device */
+	rc = register_netdev(ndev);
+	if (rc) {
+          dev_err(dev,
+                  "Cannot register network device, aborting\n");
+          goto error;
+	}
+
+	dev_info(dev, "Lowrisc Ether100MHz registered\n");
+	
+	return 0;
+
+error:
+	lowrisc_remove_ndev(ndev);
+	return rc;
+}
+
+/* Match table for OF platform binding */
+static const struct of_device_id lowrisc_100MHz_of_match[] = {
+	{ .compatible = DRIVER_NAME },
+	{ /* end of list */ },
+};
+MODULE_DEVICE_TABLE(of, lowrisc_100MHz_of_match);
+
+void lowrisc_100MHz_free(struct platform_device *of_dev)
+{
+        struct resource *iomem = platform_get_resource(of_dev, IORESOURCE_MEM, 0);
+        release_mem_region(iomem->start, resource_size(iomem));
+}
+
+int lowrisc_100MHz_unregister(struct platform_device *of_dev)
+{
+        lowrisc_100MHz_free(of_dev);
+        return 0;
+}
+
+static struct platform_driver lowrisc_100MHz_driver = {
+	.driver = {
+		.name = DRIVER_NAME,
+		.of_match_table = lowrisc_100MHz_of_match,
+	},
+	.probe = lowrisc_100MHz_probe,
+	.remove = lowrisc_100MHz_unregister,
+};
+
+module_platform_driver(lowrisc_100MHz_driver);
+
+MODULE_AUTHOR("Jonathan Kimmitt");
+MODULE_DESCRIPTION("Lowrisc Ethernet 100MHz driver");
+MODULE_LICENSE("GPL");
diff --git a/drivers/net/ethernet/lowrisc/lowrisc_100MHz.h b/drivers/net/ethernet/lowrisc/lowrisc_100MHz.h
new file mode 100644
index 000000000000..9d227c10dd36
--- /dev/null
+++ b/drivers/net/ethernet/lowrisc/lowrisc_100MHz.h
@@ -0,0 +1,55 @@
+// See LICENSE for license details.
+
+#ifndef ETH_HEADER_H
+#define ETH_HEADER_H
+
+/* Register offsets for the LowRISC Ethernet Core */
+
+/* Register offsets (in bytes) for the LowRISC Core */
+#define TXBUFF_OFFSET       0x1000          /* Transmit Buffer */
+
+#define MACLO_OFFSET        0x0800          /* MAC address low 32-bits */
+#define MACHI_OFFSET        0x0808          /* MAC address high 16-bits and MAC ctrl */
+#define TPLR_OFFSET         0x0810          /* Tx packet length */
+#define TFCS_OFFSET         0x0818          /* Tx frame check sequence register */
+#define MDIOCTRL_OFFSET     0x0820          /* MDIO Control Register */
+#define RFCS_OFFSET         0x0828          /* Rx frame check sequence register(read) and last register(write) */
+#define RSR_OFFSET          0x0830          /* Rx status and reset register */
+#define RBAD_OFFSET         0x0838          /* Rx bad frame and bad fcs register arrays */
+#define RPLR_OFFSET         0x0840          /* Rx packet length register array */
+
+#define RXBUFF_OFFSET       0x4000          /* Receive Buffer */
+
+/* MAC Ctrl Register (MACHI) Bit Masks */
+#define MACHI_MACADDR_MASK    0x0000FFFF     /* MAC high 16-bits mask */
+#define MACHI_COOKED_MASK     0x00010000     /* obsolete flag */
+#define MACHI_LOOPBACK_MASK   0x00020000     /* Rx loopback packets */
+#define MACHI_ALLPKTS_MASK    0x00400000     /* Rx all packets (promiscuous mode) */
+#define MACHI_IRQ_EN          0x00800000     /* Rx packet interrupt enable */
+
+/* MDIO Control Register Bit Masks */
+#define MDIOCTRL_MDIOCLK_MASK 0x00000001    /* MDIO Clock Mask */
+#define MDIOCTRL_MDIOOUT_MASK 0x00000002    /* MDIO Output Mask */
+#define MDIOCTRL_MDIOOEN_MASK 0x00000004    /* MDIO Output Enable Mask, 3-state enable, high=input, low=output */
+#define MDIOCTRL_MDIORST_MASK 0x00000008    /* MDIO Input Mask */
+#define MDIOCTRL_MDIOIN_MASK  0x00000008    /* MDIO Input Mask */
+
+/* Transmit Status Register (TPLR) Bit Masks */
+#define TPLR_FRAME_ADDR_MASK  0x0FFF0000     /* Tx frame address */
+#define TPLR_PACKET_LEN_MASK  0x00000FFF     /* Tx packet length */
+#define TPLR_BUSY_MASK        0x80000000     /* Tx busy mask */
+
+/* Receive Status Register (RSR) */
+#define RSR_RECV_FIRST_MASK   0x0000000F      /* first available buffer (static) */
+#define RSR_RECV_NEXT_MASK    0x000000F0      /* current rx buffer (volatile) */
+#define RSR_RECV_LAST_MASK    0x00000F00      /* last available rx buffer (static) */
+#define RSR_RECV_DONE_MASK    0x00001000      /* Rx complete */
+#define RSR_RECV_IRQ_MASK     0x00002000      /* Rx irq bit */
+
+/* General Ethernet Definitions */
+#define HEADER_OFFSET               12      /* Offset to length field */
+#define HEADER_SHIFT                16      /* Shift value for length */
+#define ARP_PACKET_SIZE             28      /* Max ARP packet size */
+#define HEADER_IP_LENGTH_OFFSET     16      /* IP Length Offset */
+
+#endif
diff --git a/drivers/spi/Kconfig b/drivers/spi/Kconfig
index 15ea11ebcbe0..86e5ffa99036 100644
--- a/drivers/spi/Kconfig
+++ b/drivers/spi/Kconfig
@@ -688,6 +688,11 @@ config SPI_OMAP24XX
 	  SPI master controller for OMAP24XX and later Multichannel SPI
 	  (McSPI) modules.
 
+config SPI_OPENTITAN_QSPI
+    tristate "OpenTitan SPI Host driver"
+    help
+      SPI host controller for the OpenTitan QSPI host
+
 config SPI_TI_QSPI
 	tristate "DRA7xxx QSPI controller support"
 	depends on ARCH_OMAP2PLUS || COMPILE_TEST
diff --git a/drivers/spi/Makefile b/drivers/spi/Makefile
index 4b34e855c841..c695999f0e64 100644
--- a/drivers/spi/Makefile
+++ b/drivers/spi/Makefile
@@ -92,6 +92,7 @@ obj-$(CONFIG_SPI_OCTEON)		+= spi-octeon.o
 obj-$(CONFIG_SPI_OMAP_UWIRE)		+= spi-omap-uwire.o
 obj-$(CONFIG_SPI_OMAP_100K)		+= spi-omap-100k.o
 obj-$(CONFIG_SPI_OMAP24XX)		+= spi-omap2-mcspi.o
+obj-$(CONFIG_SPI_OPENTITAN_QSPI)	+= spi-opentitan.o
 obj-$(CONFIG_SPI_TI_QSPI)		+= spi-ti-qspi.o
 obj-$(CONFIG_SPI_ORION)			+= spi-orion.o
 obj-$(CONFIG_SPI_PIC32)			+= spi-pic32.o
diff --git a/drivers/spi/spi-opentitan.c b/drivers/spi/spi-opentitan.c
new file mode 100644
index 000000000000..53dbb7fa2765
--- /dev/null
+++ b/drivers/spi/spi-opentitan.c
@@ -0,0 +1,755 @@
+#include <linux/io.h>
+#include <linux/interrupt.h>
+#include <linux/module.h>
+#include <linux/of.h>
+#include <linux/platform_device.h>
+#include <linux/spi/spi.h>
+
+#define MIN(a,b) ((a <= b) ? a : b)
+
+#define OPENTITAN_QSPI_DRIVER_NAME		"opentitan_qspi"
+
+// Bit 30 in Control: SW Reset
+#define OPENTITAN_QSPI_CONTROL_SPI_EN (1 << 31)
+#define OPENTITAN_QSPI_CONTROL_SW_RESET (1 << 30)
+#define OPENTITAN_QSPI_CONTROL_OUTPUT_EN (1 << 29)
+
+#define OPENTITAN_QSPI_TXFIFO_DEPTH 256
+#define OPENTITAN_QSPI_RXFIFO_DEPTH 256
+#define OPENTITAN_QSPI_FIFO_DEPTH MIN(OPENTITAN_QSPI_TXFIFO_DEPTH, OPENTITAN_QSPI_RXFIFO_DEPTH)
+
+#define OPENTITAN_QSPI_READY_TIMEOUT 10000
+#define OPENTITAN_QSPI_READ_TIMEOUT 10000
+
+// 5 us divided by 8 bits and inverted to be in integer range
+#define OPENTITAN_QSPI_IRQ_THRESHOLD 16000000
+
+#define OPENTITAN_QSPI_IP_EVENT 2
+#define OPENTITAN_QSPI_IP_ERROR 1
+
+#define SPI_XFER_BEGIN 1
+#define SPI_XFER_END 2
+
+/* opentitan qspi register set */
+enum opentitan_qspi_regs {
+	REG_INTR_STATE,		/* Interrupt State Register */
+	REG_INTR_ENABLE,	/* Interrupt Enable Register */
+	REG_INTR_TEST,		/* Interrupt Test Register */
+	REG_ALERT_TEST,		/* Alert Test Register */
+	REG_CONTROL,		/* Control Register */
+	REG_STATUS,			/* Status Register */
+	REG_CONFIGOPTS_0,	/* Configuration Options Register 1 */
+	REG_CONFIGOPTS_1,   /* Configuration Options Register 2 */
+	REG_CONFIGOPTS_2,   /* Configuration Options Register 3 */
+	REG_CSID,			/* Chip-Select ID */
+	REG_COMMAND,		/* Command Register */
+	REG_RXDATA,         /* SPI RX Data Window*/
+	REG_TXDATA,         /* SPI TX Data Window*/
+	REG_ERROR_ENABLE,	/* Controls which classes of error raise an interrupt */
+	REG_ERROR_STATUS,	/* Indicates that any errors have occured */
+	REG_EVENT_ENABLE,	/* Controls which classes of SPI events raise an interrupt */
+};
+
+/* opentitan qspi priv */
+struct opentitan_qspi_priv {
+	struct completion done;     /* wake-up from interrupt */
+	unsigned int __iomem *regs;
+  	unsigned int clk_freq; 		/* Peripheral clock frequency */
+	unsigned int max_freq; 		/* Max supported SPI frequency */
+	unsigned int cur_freq;		/* Current SPI clock frequency */
+	unsigned int cs_state; 		/* 0 = CS currently not asserted, 1 = CS currently asserted */
+	unsigned int used_cs;		/* Which CS shall be used for the current transfer */
+	unsigned int unused_cs;		/* Which CS should be used for de-selecting the target */
+	unsigned char byte_order; 	/* 1 = LSB shifted in/out first, 0 = MSB shifted in/out first */
+};
+
+
+static int opentitan_qspi_issue_dummy(struct spi_master *master, unsigned int bitlen, unsigned long flags)
+{
+	struct opentitan_qspi_priv *priv = spi_master_get_devdata(master);
+	unsigned char csaat 	= !(flags & SPI_XFER_END) && (priv->cs_state || flags & SPI_XFER_BEGIN);
+   	// Wait for the SPI host to be ready
+   	unsigned int ready_timeout = OPENTITAN_QSPI_READY_TIMEOUT;
+   	unsigned int status = 0;
+    unsigned int command = 0;
+
+	if(flags & SPI_XFER_BEGIN){
+		priv->cs_state = 1;
+		iowrite32(priv->used_cs, priv->regs + REG_CSID);
+	}
+
+	// Just setting the CS
+	if(bitlen == 0){
+		if(flags & SPI_XFER_END){
+			priv->cs_state = 0;
+			iowrite32(priv->unused_cs, priv->regs + REG_CSID);
+			opentitan_qspi_issue_dummy(master, 8, 0);
+		}
+		return 0;
+	}
+
+    do {
+    	status = ioread32((volatile unsigned int *) (priv->regs + REG_STATUS));
+       	ready_timeout--;
+    } while(!(status >> 31) && ready_timeout > 0);
+
+    if(ready_timeout == 0 && !(status >> 31)){
+       	dev_err(&master->dev, "opentitan_qspi_issue_dummy: Ready did not assert. Status: 0x%x\r\n", status);
+       	return -1;
+    }
+
+	command = ((bitlen & 0x1FF) - 1) | ((csaat & 0x1) << 9);
+	iowrite32(command, priv->regs + REG_COMMAND);
+
+	do {
+		status = ioread32(priv->regs + REG_STATUS);
+	} while((status >> 30) & 0x1);
+
+	if(flags & SPI_XFER_END){
+		priv->cs_state = 0;
+		iowrite32(priv->unused_cs, priv->regs + REG_CSID);
+	}
+
+	return 0;
+}
+
+// Expects the FIFOs to be empty and returns once the FIFOs are empty again
+static int opentitan_qspi_xfer_single(struct spi_master *master, unsigned int bitlen,
+				      const void *dout, void *din, unsigned long flags)
+{
+	struct opentitan_qspi_priv *priv = spi_master_get_devdata(master);
+	unsigned int num_bytes 	= bitlen/8;
+	unsigned int command = 0;
+    unsigned int status  = 0;
+    unsigned int ready_timeout = OPENTITAN_QSPI_READY_TIMEOUT;
+	unsigned char csaat 	= !(flags & SPI_XFER_END) && (priv->cs_state || flags & SPI_XFER_BEGIN);
+	unsigned char dir   	= (din != NULL) | ((dout != NULL) << 1);
+
+    if(!dout && !din)
+		return opentitan_qspi_issue_dummy(master, bitlen, flags);
+
+	if(bitlen % 8 != 0){
+		dev_err(&master->dev, "Transfers must be multiples of 8 bit long\n");
+		return -EINVAL;
+	}
+	
+	if(flags & SPI_XFER_BEGIN){
+		priv->cs_state = 1;
+		iowrite32(priv->used_cs, priv->regs + REG_CSID);
+	}
+
+	// Just setting the CS
+	if(bitlen == 0){
+		if(flags & SPI_XFER_END){
+			priv->cs_state = 0;
+			iowrite32(priv->unused_cs, priv->regs + REG_CSID);
+			opentitan_qspi_issue_dummy(master, 8, 0);
+		}
+		return 0;
+	}
+		
+	if(dir >> 1){
+	    unsigned int i = 0;
+		// Take care of the word aligned part
+		for(; i < num_bytes/4; i++){
+			unsigned char tmp[4];
+
+			if(!priv->byte_order){
+				tmp[3] = ((unsigned char *) dout)[4*i];
+				tmp[2] = ((unsigned char *) dout)[4*i + 1];
+				tmp[1] = ((unsigned char *) dout)[4*i + 2];
+				tmp[0] = ((unsigned char *) dout)[4*i + 3];
+			} else {
+				// Read from dout according to its alignment
+               	// 4 byte
+                if(!((long int) dout & 0x3L)){
+                	*((unsigned int *) tmp) = *((unsigned int *) (dout+4*i));
+ 
+                // 2 byte
+                } else if (!((long int) dout & 0x1L)){
+                	*((unsigned short *) tmp)     = *((unsigned short *) (dout+4*i));
+                	*((unsigned short *) (tmp+2)) = *((unsigned short *) (dout+4*i+2));
+ 
+                // 1 byte
+                } else {
+                	tmp[0] = ((unsigned char *) dout)[4*i];
+                	tmp[1] = ((unsigned char *) dout)[4*i + 1];
+                	tmp[2] = ((unsigned char *) dout)[4*i + 2];
+                	tmp[3] = ((unsigned char *) dout)[4*i + 3];
+                }
+			}
+
+			iowrite32(*((unsigned int *) tmp), priv->regs + REG_TXDATA);
+		}
+
+		// Less than a full word left
+		if(i*4 < num_bytes){
+			unsigned char tmp[4];
+
+			if(!priv->byte_order){
+				// We are in here so at least one byte remains
+				tmp[3] = ((unsigned char *) dout)[i*4];
+				tmp[2] = ((num_bytes - i*4) >= 2) ? ((unsigned char *) dout)[i*4 + 1] : 0;
+				tmp[1] = ((num_bytes - i*4) == 3) ? ((unsigned char *) dout)[i*4 + 2] : 0;
+			
+				// Cannot need filling as it would have been taken care of by the loop then
+				tmp[0] = 0;
+			} else {
+				// We are in here so at least one byte remains
+				tmp[0] = ((unsigned char *) dout)[i*4];
+				tmp[1] = ((num_bytes - i*4) >= 2) ? ((unsigned char *) dout)[i*4 + 1] : 0;
+				tmp[2] = ((num_bytes - i*4) == 3) ? ((unsigned char *) dout)[i*4 + 2] : 0;
+			
+				// Cannot need filling as it would have been taken care of by the loop then
+				tmp[3] = 0;
+			}
+			
+			iowrite32(*((unsigned int *) tmp), priv->regs + REG_TXDATA);
+		}
+	}
+
+	// Set the correct transfer mode
+	command = ((num_bytes & 0x1FF) - 1) | ((csaat & 0x1) << 9) | (dir << 12);
+
+    // Wait for the SPI host to be ready
+    do {
+       	status = ioread32((volatile unsigned int *) (priv->regs + REG_STATUS));
+       	ready_timeout--;
+    } while(!(status >> 31) && ready_timeout > 0);
+
+    if(ready_timeout == 0 && !(status >> 31)){
+       	dev_err(&master->dev, "opentitan_qspi_xfer_single: Ready did not assert. Status: 0x%x\r\n", status);
+       	return -1;
+    }
+
+	// Enable the interrupt to wait for completion of the transfer
+	if((num_bytes > 1) && ((priv->cur_freq/num_bytes) <= OPENTITAN_QSPI_IRQ_THRESHOLD)){
+		reinit_completion(&priv->done);
+
+		// Disable event interrupts
+		iowrite32(0x1, priv->regs + REG_INTR_ENABLE);
+
+		// Enable idle event
+		iowrite32(0x20, priv->regs + REG_EVENT_ENABLE);
+
+		// Re-enable event interrupts
+		iowrite32(0x3, priv->regs + REG_INTR_ENABLE);
+	}
+
+	// Start transaction by writing to the command register
+	iowrite32(command, priv->regs + REG_COMMAND);
+
+	if((num_bytes > 1) && ((priv->cur_freq/num_bytes) <= OPENTITAN_QSPI_IRQ_THRESHOLD)){
+		wait_for_completion(&priv->done);
+	}
+
+	// Wait for the FIFOs to be empty (full) if we had an actual data transfer
+	if(dir > 0){
+		status = 0;
+
+		// RX only or RX/TX
+		if(dir == 1 || dir == 3) {
+			unsigned int bytes_rcvd = 0;
+			do {
+				status = ioread32((volatile unsigned int *) (priv->regs + REG_STATUS));
+
+				if((status >> 8) & 0xFF){
+					if(bytes_rcvd < num_bytes){
+						unsigned char *dst = (unsigned char *) din;
+						unsigned int word = ioread32(priv->regs + REG_RXDATA);
+
+						if((num_bytes - bytes_rcvd) >= 4){
+							if(!priv->byte_order){
+								dst[3] =  word        & 0xFF;
+								dst[2] = (word >>  8) & 0xFF;
+								dst[1] = (word >> 16) & 0xFF;
+								dst[0] = (word >> 24) & 0xFF;
+							} else {
+								// Store received data into din according to it's alignment
+                            	// 4 byte
+                                if(!((long int) din & 0x3L)){
+                                	*((unsigned int *) din) = word;
+ 
+                                // 2 byte
+                                } else if (!((long int) din & 0x1L)) {
+                                	*((unsigned short *) din)   =  word        & 0xFFFF;
+                                	*((unsigned short *) din+2) = (word >> 16) & 0xFFFF;
+ 
+                                // 1 byte
+                                } else {
+                                	dst[0] =  word        & 0xFF;
+                                	dst[1] = (word >>  8) & 0xFF;
+                                	dst[2] = (word >> 16) & 0xFF;
+                                	dst[3] = (word >> 24) & 0xFF;
+                                }
+							}
+
+							din += 4;
+							bytes_rcvd += 4;
+						} else {
+							if(!priv->byte_order){
+								// We are in here so at least one byte remains
+								dst[0] = (word >> 24) & 0xFF;
+								bytes_rcvd++;
+
+								if((num_bytes - bytes_rcvd) >= 1){
+									dst[1] = (word >> 16) & 0xFF;
+									bytes_rcvd++;
+								}
+
+								if((num_bytes - bytes_rcvd) == 1){
+									dst[2] = (word >> 8) & 0xFF;
+									bytes_rcvd++;
+								}
+
+							} else {
+								// We are in here so at least one byte remains
+								dst[0] = word & 0xFF;
+                                bytes_rcvd++;
+
+								if((num_bytes - bytes_rcvd) >= 1){
+									dst[1] = (word >> 8) & 0xFF;
+									bytes_rcvd++;
+								}
+
+								if((num_bytes - bytes_rcvd) == 1){
+									dst[2] = (word >> 16) & 0xFF;
+									bytes_rcvd++;
+								}
+							}
+						}
+
+					// Somehow we have too much data??
+					} else {
+						(void) ioread32(priv->regs + REG_RXDATA);
+						dev_err(&master->dev, "Device returned more data than we requested. num_bytes = 0x%x, bytes_rcvd = 0x%x, status = 0x%x\n",
+								num_bytes, bytes_rcvd, status);
+					}
+				}
+			} while(((status >> 8) & 0xFF) || ((status >> 30) & 0x1));
+
+            // Wait for the last bytes of the transfer
+			if(bytes_rcvd < num_bytes){
+                unsigned int read_timeout = OPENTITAN_QSPI_READ_TIMEOUT;
+                unsigned int word;
+				unsigned char *dst = (unsigned char *) din;
+
+                do {
+                	status = ioread32((volatile unsigned int *) (priv->regs + REG_STATUS));
+                	read_timeout--;
+				} while(!((status >> 8) & 0xFF) && read_timeout > 0);
+
+                if(read_timeout == 0 && !((status >> 8) & 0xFF)){
+                	dev_err(&master->dev, "RX queue did not notify us about the last bytes\r\n");
+                	return -1;
+                }
+        
+				word = ioread32((volatile unsigned int *) (priv->regs + REG_RXDATA));
+
+				if(!priv->byte_order){
+					// We are in here so at least one byte remains
+					dst[0] = (word >> 24) & 0xFF;
+					bytes_rcvd++;
+
+					if((num_bytes - bytes_rcvd) >= 1){
+						dst[1] = (word >> 16) & 0xFF;
+						bytes_rcvd++;
+					}
+
+					if((num_bytes - bytes_rcvd) >= 1){
+						dst[2] = (word >> 8) & 0xFF;
+						bytes_rcvd++;
+					}
+
+					if((num_bytes - bytes_rcvd) == 1){
+                        dst[3] = word & 0xFF;
+						bytes_rcvd++;
+					}
+
+				} else {
+					// We are in here so at least one byte remains
+					dst[0] = word & 0xFF;
+                    			bytes_rcvd++;
+
+					if((num_bytes - bytes_rcvd) >= 1){
+						dst[1] = (word >> 8) & 0xFF;
+						bytes_rcvd++;
+					}
+
+					if((num_bytes - bytes_rcvd) >= 1){
+						dst[2] = (word >> 16) & 0xFF;
+						bytes_rcvd++;
+					}
+
+					if((num_bytes - bytes_rcvd) == 1){
+                        dst[3] = (word >> 24) & 0xFF;
+						bytes_rcvd++;
+					}
+				}
+			}
+
+		// TX Only
+		} else if(dir == 2) {
+			// Wait for SPI host do go idle
+			do {
+				status = ioread32(priv->regs + REG_STATUS);
+			} while((status >> 30) & 0x1);
+ 		
+		// What mode is this??
+		} else {
+			dev_err(&master->dev, "This direction is unknown: %d\n", dir);
+			return -EINVAL;
+		}
+	}
+
+	if(flags & SPI_XFER_END){
+		priv->cs_state = 0;
+		iowrite32(priv->unused_cs, priv->regs + REG_CSID);
+	}
+
+	return 0;
+}
+
+static int opentitan_qspi_xfer(struct spi_master *master, unsigned int bitlen,
+			       const void *dout, void *din, unsigned long flags)
+{
+    //dev_info(&master->dev, "opentitan_qspi_xfer: bitlen = 0x%x, dout = 0x%x, din = 0x%x, flags = %lu\r\n",
+    //         bitlen, (unsigned int) dout, (unsigned int) din, flags);
+
+	// Yay a single transaction
+	if(bitlen <= OPENTITAN_QSPI_FIFO_DEPTH*8){
+		return opentitan_qspi_xfer_single(master, bitlen, dout, din, flags);
+
+	// Aww multiple transactions
+	} else {
+		unsigned long first_flags = flags & SPI_XFER_BEGIN;
+		unsigned long last_flags  = flags & SPI_XFER_END;
+		unsigned int  num_txns    = (bitlen + OPENTITAN_QSPI_FIFO_DEPTH*8 - 1)/(OPENTITAN_QSPI_FIFO_DEPTH*8);
+        unsigned int i;
+		
+		for(i = 0; i < num_txns; i++){
+			unsigned long flags = (i == 0) ? first_flags :
+					      		  (i == num_txns-1) ? last_flags  : 0;
+			unsigned int ret = 0;
+			unsigned int len = ((bitlen - i*OPENTITAN_QSPI_FIFO_DEPTH*8) < OPENTITAN_QSPI_FIFO_DEPTH*8) ?
+					    		(bitlen - i*OPENTITAN_QSPI_FIFO_DEPTH*8) : OPENTITAN_QSPI_FIFO_DEPTH*8;
+			void const *out = NULL;
+			void *in  = NULL;
+
+			if(dout)
+				out = (void *) (dout + i*OPENTITAN_QSPI_FIFO_DEPTH);
+
+			if(din)
+				in  = (void *) (din + i*OPENTITAN_QSPI_FIFO_DEPTH);
+
+			ret = opentitan_qspi_xfer_single(master, len, out, in, flags);
+			
+			if(ret)
+				return ret;
+		}
+
+		return 0;
+	}
+}
+
+static int opentitan_qspi_set_speed(struct spi_master *master, uint speed)
+{
+	unsigned long int clkdiv = 0;
+	u32 configopts = 0;
+	struct opentitan_qspi_priv *priv = spi_master_get_devdata(master);
+
+	if(speed > priv->max_freq){
+		dev_info(&master->dev, "Requested frequency is higher than maximum possible frequency!\n");
+		dev_info(&master->dev, "Req: %d, Max: %d\n", speed, priv->max_freq);
+		speed = priv->max_freq;
+	}
+
+	priv->cur_freq = speed;
+
+	// SPI_CLK = SYS_CLK/(2*(clkdiv+1))
+	// clkdiv = SYS_CLK/(2*SPI_CLK) - 1
+	clkdiv = priv->clk_freq + 2*speed - 1L;
+	clkdiv = clkdiv/(2*speed) - 1L;
+
+	if(clkdiv != (clkdiv & (~(-1 << 16)))){
+		dev_info(&master->dev, "Calculated clock divider overflows the hardware register! Using maximum value\n");
+		clkdiv = ~(-1 << 16);
+	}
+
+	configopts = (u32) ioread32(priv->regs + REG_CONFIGOPTS_0 + priv->used_cs);
+	configopts = (configopts & (-1 << 16)) | (clkdiv & ~(-1 << 16));
+	iowrite32(configopts, priv->regs + REG_CONFIGOPTS_0 + priv->used_cs);
+
+	// This is dirty... we are wasting a whole chip select just to be able to control the chipselect
+	// independently of the rest of the SPI bus
+	iowrite32(configopts, priv->regs + REG_CONFIGOPTS_0 + priv->unused_cs);
+
+	return 0;
+}
+
+static int opentitan_qspi_set_mode(struct spi_master *master, uint mode)
+{
+	struct opentitan_qspi_priv *priv = spi_master_get_devdata(master);
+	unsigned int configopts = 0;
+
+	configopts = (unsigned int) ioread32(priv->regs + REG_CONFIGOPTS_0 + priv->used_cs);
+	configopts = (configopts & 0xFFFF) | (0xFFF << 16) | ((mode & 0x3) << 30);
+	iowrite32(configopts, priv->regs + REG_CONFIGOPTS_0 + priv->used_cs);
+	iowrite32(configopts, priv->regs + REG_CONFIGOPTS_0 + priv->unused_cs);
+
+	return 0;
+}
+
+static irqreturn_t opentitan_qspi_irq(int irq, void *dev_id)
+{
+	struct opentitan_qspi_priv *priv = (struct opentitan_qspi_priv *) dev_id;
+	unsigned int ie = 0;
+	unsigned int ip = ioread32(priv->regs + REG_INTR_STATE);
+	unsigned int status = ioread32(priv->regs + REG_STATUS);
+
+	if (ip & OPENTITAN_QSPI_IP_EVENT){
+		if(!((status >> 30) & 0x1)){
+			unsigned int intr_enable = ioread32(priv->regs + REG_INTR_ENABLE);
+
+			// Disable event interrupts
+			iowrite32(intr_enable & 0x1, priv->regs + REG_INTR_ENABLE);
+
+			// Disable all events
+			iowrite32(0x0, priv->regs + REG_EVENT_ENABLE);
+
+			// Clear event interrupt flag
+			iowrite32(0x2, priv->regs + REG_INTR_STATE);
+
+			complete(&priv->done);
+			return IRQ_HANDLED;
+		}
+	} else if (ip & OPENTITAN_QSPI_IP_ERROR) {
+		unsigned int err_status = ioread32(priv->regs + REG_ERROR_STATUS);
+
+		dev_err(NULL, "opentitan_qspi: error interrupt! error_status = 0x%x\r\n", err_status);
+		
+		// Badly handle the error by just clearing the flags
+		iowrite32(0x1F, priv->regs + REG_ERROR_STATUS);
+		iowrite32(0x1, priv->regs + REG_INTR_STATE);
+
+		complete(&priv->done);
+		return IRQ_HANDLED;
+	}
+	return IRQ_NONE;
+}
+
+static int opentitan_qspi_prepare_message(struct spi_master *master, struct spi_message *msg)
+{
+	struct opentitan_qspi_priv *priv = spi_master_get_devdata(master);
+	struct spi_device *device = msg->spi;
+
+	// Select correct chip select
+	priv->used_cs = device->chip_select;
+
+	if (device->mode & SPI_CS_HIGH)
+		dev_err(&master->dev, "SPI_CS_HIGH requested!\r\n");
+
+	// Set the correct mode
+	return opentitan_qspi_set_mode(master, device->mode & 0x3);
+}
+
+static void opentitan_qspi_set_cs(struct spi_device *device, bool is_high)
+{
+	struct opentitan_qspi_priv *priv = spi_master_get_devdata(device->master);
+
+	// is_high is the logic level of the CS
+	if(!is_high) {
+		priv->cs_state = 1;
+		iowrite32(priv->used_cs, priv->regs + REG_CSID);
+	} else {
+		priv->cs_state = 0;
+		iowrite32(priv->unused_cs, priv->regs + REG_CSID);
+		opentitan_qspi_issue_dummy(device->master, 8, 0);
+	}
+}
+
+static int opentitan_qspi_transfer_one(struct spi_master *master, struct spi_device *device,
+									   struct spi_transfer *t)
+{
+	struct opentitan_qspi_priv *priv = spi_master_get_devdata(master);
+
+	/* Set the speed according to the transfer or use the last speed */
+	if (t->speed_hz){
+		opentitan_qspi_set_speed(master, t->speed_hz);
+		priv->cur_freq = t->speed_hz;
+	} else {
+		opentitan_qspi_set_speed(master, priv->cur_freq);
+	}
+	
+	return opentitan_qspi_xfer(master, t->len*8, t->tx_buf, t->rx_buf, 0);
+}
+
+static int opentitan_qspi_probe(struct platform_device *pdev)
+{
+    struct opentitan_qspi_priv *priv;
+	struct spi_master *master;
+	u32 status = 0, loop_count = 0, ret = 0, err_irq = 0, evt_irq = 0;
+
+	master = spi_alloc_master(&pdev->dev, sizeof(struct opentitan_qspi_priv));
+	if (!master) {
+		dev_err(&pdev->dev, "out of memory\r\n");
+		return -ENOMEM;
+	}
+
+	priv = spi_master_get_devdata(master);
+	init_completion(&priv->done);
+	platform_set_drvdata(pdev, master);
+	
+	priv->regs = devm_platform_ioremap_resource(pdev, 0);
+	if (IS_ERR(priv->regs)) {
+		ret = PTR_ERR(priv->regs);
+		goto put_master;
+	}
+
+	// Get the error interrupt
+	err_irq = platform_get_irq(pdev, 0);
+	if (err_irq < 0) {
+		ret = err_irq;
+		goto put_master;
+	}
+
+	// Get the event interrupt
+	evt_irq = platform_get_irq(pdev, 1);
+	if (evt_irq < 0) {
+		ret = evt_irq;
+		goto put_master;
+	}
+
+	ret = of_property_read_u32(pdev->dev.of_node, "clock-frequency",
+	                           &priv->clk_freq);
+	if(ret < 0){
+		dev_err(&pdev->dev, "Unable to determine SPI clock frequency\r\n");
+		goto put_master;
+	}
+
+	ret = of_property_read_u32(pdev->dev.of_node, "max-frequency",
+							   &priv->max_freq);
+	if(ret < 0)
+		priv->max_freq = 400000;	// Default to 400 kHz
+
+	// Use first CS by default
+	priv->unused_cs = 2;
+    priv->used_cs = 0;
+
+	// Read the byte order
+	status = ioread32(priv->regs + REG_STATUS);
+	priv->byte_order = (status >> 22) & 0x1;
+
+	/* Define our master */
+	master->dev.of_node 		= pdev->dev.of_node;
+	master->bus_num 			= pdev->id;
+	master->num_chipselect 		= 1;
+	master->mode_bits 			= SPI_CPHA | SPI_CPOL; // | SPI_CS_HIGH;
+	master->bits_per_word_mask 	= SPI_BPW_MASK(8);
+	master->max_speed_hz		= priv->max_freq;
+	master->flags 				= 0;
+	master->prepare_message		= opentitan_qspi_prepare_message;
+	master->set_cs				= opentitan_qspi_set_cs;
+	master->transfer_one		= opentitan_qspi_transfer_one;
+
+	pdev->dev.dma_mask = NULL; 
+
+	// Disable all interrupts
+	iowrite32(0, priv->regs + REG_INTR_ENABLE);
+	iowrite32(0, priv->regs + REG_EVENT_ENABLE);
+
+	// Assert SW reset of the SPI Host
+	iowrite32(OPENTITAN_QSPI_CONTROL_SW_RESET, priv->regs + REG_CONTROL);
+
+	// Wait until the FIFOs are drained
+	do {
+		status = (int) ioread32(priv->regs + REG_STATUS);
+		loop_count++;
+
+		if(loop_count >= 1000000){
+			iowrite32(0, priv->regs + REG_CONTROL);
+			return -EINVAL;
+		}
+	} while ((status >> 30) & 1 || (status << 16));
+
+	// Deassert SW reset and assert enable signal => Start SPI Host
+	iowrite32((1 << 31), priv->regs + REG_CONTROL);
+
+	// Configure the CS
+	// De-select the connected peripheral by default
+	iowrite32(priv->unused_cs, priv->regs + REG_CSID);
+
+	// Enable the SPI and its output
+  	iowrite32(OPENTITAN_QSPI_CONTROL_OUTPUT_EN  | OPENTITAN_QSPI_CONTROL_SPI_EN, priv->regs + REG_CONTROL);
+
+	/* Register for SPI event interrupt */
+	ret = devm_request_irq(&pdev->dev, evt_irq, opentitan_qspi_irq, 0,
+			       dev_name(&pdev->dev), priv);
+	if (ret) {
+		dev_err(&pdev->dev, "Unable to bind to event interrupt\n");
+		goto put_master;
+	}
+
+	/* Register for SPI error interrupt */
+	ret = devm_request_irq(&pdev->dev, err_irq, opentitan_qspi_irq, 0,
+			       dev_name(&pdev->dev), priv);
+	if (ret) {
+		dev_err(&pdev->dev, "Unable to bind to error interrupt\n");
+		goto put_master;
+	}
+
+	ret = devm_spi_register_master(&pdev->dev, master);
+	if(ret < 0){
+		dev_err(&pdev->dev, "spi_register_master failed\r\n");
+		goto put_master;
+	}
+
+	// Enable all errors and the error interrupt
+	iowrite32(0x1F, priv->regs + REG_ERROR_ENABLE);
+	iowrite32(0x1, priv->regs + REG_INTR_ENABLE);
+
+	dev_info(&pdev->dev, "initialized; err_irq = %d, evt_irq = %d, num cs = %d, byte order = %d\r\n",
+			 err_irq, evt_irq, priv->unused_cs, priv->byte_order);
+    dev_info(&pdev->dev, "Params: regs = 0x%x, clk_freq = %u, max_freq = %u, cs_state = %u\r\n",
+             (unsigned int) priv->regs, priv->clk_freq, priv->max_freq, priv->cs_state);
+    dev_info(&pdev->dev, "Params: used_cs = %u, unused_cs = %u\r\n", priv->used_cs, priv->unused_cs);
+    dev_info(&pdev->dev, "Status reg @ 0x%x: 0x%x\r\n", (unsigned int) (priv->regs + REG_STATUS), ioread32(priv->regs + REG_STATUS));
+
+	return 0;
+
+put_master:
+	spi_master_put(master);
+
+	return ret;
+}
+
+static void opentitan_qspi_remove(struct platform_device *pdev)
+{
+	struct spi_master *master = platform_get_drvdata(pdev);
+	struct opentitan_qspi_priv *priv = spi_master_get_devdata(master);
+
+	// Disable all interrupts
+	iowrite32(0, priv->regs + REG_INTR_ENABLE);
+	iowrite32(0, priv->regs + REG_EVENT_ENABLE);
+}
+
+static const struct of_device_id opentitan_qspi_of_match[] = {
+	{ .compatible = "opentitan,spi-host" },
+	{ }
+};
+
+MODULE_DEVICE_TABLE(of, opentitan_qspi_of_match);
+
+static struct platform_driver opentitan_qspi_driver = {
+	.probe = opentitan_qspi_probe,
+	.remove_new = opentitan_qspi_remove,
+	.driver = {
+		.name = OPENTITAN_QSPI_DRIVER_NAME,
+		.of_match_table = opentitan_qspi_of_match
+	},
+};
+module_platform_driver(opentitan_qspi_driver);
+
+MODULE_AUTHOR("PULP Platform <info@pulp-platform.org");
+MODULE_DESCRIPTION("OpenTitan SPI Host driver");
+MODULE_LICENSE("GPL");
\ No newline at end of file
-- 
2.34.1

